homeassistant/components/hardware/websocket_api.py

```diff
@@ -45,3 +64,51 @@ async def ws_info(
                 hardware_info.append(asdict(platform.async_info(hass)))
 
     connection.send_result(msg["id"], {"hardware": hardware_info})
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "hardware/subscribe_system_status",
+    }
+)
+@websocket_api.async_response
+async def ws_subscribe_system_status(hass, connection, msg):
+    """Subscribe to system status updates."""
+
+    system_status: SystemStatus = hass.data[DOMAIN]["system_status"]
+
+    @callback
+    def async_update_status(now: datetime) -> None:
+        cpu_percentage = round(
+            system_status.ha_psutil.psutil.cpu_percent(interval=None)
```

Comment:
Are the psutil calls async safe?

Reply:
The calls are opening and reading files in the `/proc` virtual file system, I'm not sure if such calls can block?

Reply:
Reading proc can block

Reply:
@agners thought it wouldn't block, at least not for the files our calls will read from; please see comments in the code:
```py
    @callback
    def async_update_status(now: datetime) -> None:
        # Although cpu_percent and virtual_memory access files in the /proc vfs, those
        # accesses do not block and we don't need to wrap the calls in an executor.
        # https://elixir.bootlin.com/linux/v5.19.4/source/fs/proc/stat.c
        # https://elixir.bootlin.com/linux/v5.19.4/source/fs/proc/meminfo.c#L32
```

Do you think that's not the case @bdraco ?

Reply:
Yes, reading from /proc can block. But the two particular files we are reading `/proc/stat` and `/proc/meminfo` seem not to, looking at the source:
https://elixir.bootlin.com/linux/v5.15.63/source/fs/proc/meminfo.c#L32
https://elixir.bootlin.com/linux/v5.15.63/source/fs/proc/stat.c#L108

There seem to be no sleep in these functions or functions it calls. The data is simply printed from current variables.

The only return value is 0 (success).